File: DSA/src/Questions/String/LongestSubsequentSequence.java
package Questions.String;

import java.util.HashMap;

public class LongestSubsequentSequence {
    public int traverse(HashMap<Integer, Boolean> hash, int num, int op) {
        // System.out.println(num + " " + op);
        // hash.forEach( (k, v) -> System.out.print(k + " " + v + " || "));
        // System.out.println("");
        int count = 0;
        while (hash.getOrDefault(num, false)) {
            hash.put(num, false);
            num = num + op;
            count++;
        }
        return count;
    }

    public int longestConsecutive(int[] nums) {
        HashMap<Integer, Boolean> hash = new HashMap<>();

        for (int num : nums) {
            hash.put(num, true);
        }

        int result = 0;
        for (int num : nums) {
            int curResult = traverse(hash, num + 1, 1) + traverse(hash, num, -1);
            // System.out.println(curResult + " " + num   + " " + result);
            result = Math.max(result, curResult);
        }
        return result;
    }

    public int longestConsecutiveOptimized(int[] nums) {
        HashMap<Integer, Boolean> hash = new HashMap<>();
        for (int num : nums) {
            hash.put(num, true);
        }

        int result = 0;
        for (int num : nums) {
            if (!hash.getOrDefault(num - 1, false)) {
                int curResult = traverse(hash, num + 1, 1);
                result = Math.max(result, curResult + 1);
            }
        }
        return result;
    }
    /*
     * Time complexity: O(n) Space complexity: O(n)
     */

    public static void main(String[] args) {
        int[] nums = new int[] { 100, 4, 200, 1, 3, 2 };
        LongestSubsequentSequence longestSubsequentSequence = new LongestSubsequentSequence();
        System.out.println(longestSubsequentSequence.longestConsecutive(nums));
        System.out.println(longestSubsequentSequence.longestConsecutiveOptimized(nums));
    }
}

File: DSA/Python/Questions/String/LongestSubsequentSequence.py
"""
# Question: Longest Consecutive Sequence
# Link: https://leetcode.com/problems/longest-consecutive-sequence/

# Time Complexity: O(n)
# Space Complexity: O(n)

# Algorithm:
# 1. Use hashmap to track numbers
# 2. Find sequence start points
# 3. Count sequence lengths
# 4. Return maximum sequence length

# Key Components:
# - longest_consecutive(): Main implementation
# - longest_consecutive_optimized(): Optimized version
# - traverse(): Helper for sequence counting
"""


class LongestSubsequentSequence:
    def traverse(self, hash_map: dict, num: int, op: int) -> int:
        count = 0
        while hash_map.get(num, False):
            hash_map[num] = False
            num = num + op
            count += 1
        return count

    def longest_consecutive(self, nums: list[int]) -> int:
        hash_map = {num: True for num in nums}
        result = 0

        for num in nums:
            cur_result = self.traverse(hash_map, num + 1, 1) + self.traverse(
                hash_map, num, -1
            )
            result = max(result, cur_result)

        return result

    def longest_consecutive_optimized(self, nums: list[int]) -> int:
        hash_map = {num: True for num in nums}
        result = 0

        for num in nums:
            if not hash_map.get(num - 1, False):
                cur_result = self.traverse(hash_map, num + 1, 1)
                result = max(result, cur_result + 1)

        return result


def main():
    solution = LongestSubsequentSequence()
    nums = [100, 4, 200, 1, 3, 2]
    print(solution.longest_consecutive(nums))
    print(solution.longest_consecutive_optimized(nums))


if __name__ == "__main__":
    main()
